src/pgwire-common/src/conn.rs

```diff
@@ -117,3 +119,147 @@ where
         }
     }
 }
+
+/// Metadata about a user that is required to allocate a [`ConnectionHandle`].
+#[derive(Debug, Clone, Copy)]
+pub struct UserMetadata {
+    pub is_admin: bool,
+    pub should_limit_connections: bool,
+}
+
+#[derive(Debug, Clone)]
+pub struct ConnectionCounter {
+    inner: Arc<Mutex<ConnectionCounterInner>>,
+}
+
+impl ConnectionCounter {
+    /// Returns a [`ConnectionHandle`] which must be kept alive for the entire duration of the
+    /// external connection.
+    ///
+    /// Dropping the [`ConnectionHandle`] decrements the connection count.
+    pub fn allocate_connection(
+        &self,
+        metadata: impl Into<UserMetadata>,
+    ) -> Result<Option<ConnectionHandle>, ConnectionError> {
+        let mut inner = self.inner.lock().expect("environmentd panicked");
+        let metadata = metadata.into();
+
+        if !metadata.should_limit_connections {
+            return Ok(None);
+        }
+
+        if (metadata.is_admin && inner.reserved_remaining() > 0)
+            || inner.non_reserved_remaining() > 0
+        {
+            inner.inc_connection_count();
+            Ok(Some(self.create_handle()))
+        } else {
+            Err(ConnectionError::TooManyConnections {
+                current: inner.current,
+                limit: inner.limit,
+            })
+        }
+    }
+
+    /// Updates the maximum number of connections we allow.
+    pub fn update_limit(&self, new_limit: u64) {
+        let mut inner = self.inner.lock().expect("environmentd panicked");
+        inner.limit = new_limit;
+    }
+
+    /// Updates the number of connections we reserve for superusers.
+    pub fn update_superuser_reserved(&self, new_reserve: u64) {
+        let mut inner = self.inner.lock().expect("environmentd panicked");
+        inner.superuser_reserved = new_reserve;
+    }
+
+    fn create_handle(&self) -> ConnectionHandle {
+        let inner = Arc::clone(&self.inner);
+        let decrement_fn = Box::new(move || {
+            let mut inner = inner.lock().expect("environmentd panicked");
+            inner.dec_connection_count();
+        });
+
+        ConnectionHandle {
+            decrement_fn: Some(decrement_fn),
+        }
+    }
+}
+
+impl Default for ConnectionCounter {
+    fn default() -> Self {
+        let inner = ConnectionCounterInner::new(10, 3);
+        ConnectionCounter {
+            inner: Arc::new(Mutex::new(inner)),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ConnectionCounterInner {
+    /// Current number of connections.
+    current: u64,
+    /// Total number of connections allowed.
+    limit: u64,
+    /// Number of connections we'll reserve for super users.
```

Comment:
It took me awhile to figure out that this is the number of connections in `limit`. Maybe something like
```suggestion
    /// Number of connections in `limit` we'll reserve for super users.
```
would make it clearer.

Reply:
Updated!